### PR TITLE
Add functions for Base64 conversions.

### PR DESCRIPTION
**Pull request by antkmsft**
Maps to https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/identity/Azure.Identity/src/Base64Url.cs

### CODE REVIEW

File: sdk/core/core/src/az_base64.c [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+#include <az_base64.h>
+#include <az_contract.h>
+
+#include <assert.h>
+
+#include <_az_cfg.h>
+
+enum {
+  TRIBYTE_OCTETS = 3,
```
Comment (sergey-shandar) [inline sdk/core/core/src/az_base64.c]: AZ_TRIBYTE_...

File: sdk/core/core/test/main.c [hunk 2]
```diff
@@ -424,5 +425,105 @@ int main() {
 
     TEST_ASSERT(az_const_span_eq(az_span_to_const_span(actual), expected));
   }
+  {
+    az_const_span const decoded0 = AZ_STR("");
```
Comment (sergey-shandar) [inline sdk/core/core/test/main.c]: IMHO these constants should be global and static.

File: sdk/core/core/test/main.c [hunk 3]
```diff
@@ -424,5 +425,105 @@ int main() {
...
+    az_const_span const encoded5u = AZ_STR("MTIzNDU");
+    az_const_span const encoded6u = AZ_STR("MTIzNDU2");
+
+    az_const_span const encoded_bin1
+        = AZ_STR("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/");
+
+    az_const_span const encoded_bin1u
+        = AZ_STR("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_");
+
+    uint8_t const decoded_bin1_buf[]
```
Comment (sergey-shandar) [inline sdk/core/core/test/main.c]: same here

File: sdk/core/core/src/az_uri.c [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  }
+
+  p[0] = '%';
+  p[1] = int4_to_hex(c >> 4);
+  p[2] = int4_to_hex(c & 0x0F);
+
+  return 3;
+}
+
+az_result az_uri_encode(
```
Comment (vhvb1989) [inline sdk/core/core/src/az_uri.c]: some comments / examples would be awesome

**Comment by barcharcraz** [comment]
Looks good, however it's very likely any TLS implementation will have functions to do base64 conversions already.

There's an argument for writing our own though I suppose.

**Comment by antkmsft** [comment]
@barcharcraz,
> There's an argument for writing our own though I suppose.

We can remove it later if we find a good one. For now, we have libcurl, for example, and it allocates.
> _"Given a base64 NUL-terminated string at src, decode it and return a pointer in *outptr to a newly allocated memory area holding decoded data. Size of decoded data is returned in variable pointed by outlen"._
 - https://github.com/curl/curl/blob/d81dbae19f8876ad472e445d89760970c79cceaa/lib/base64.c#L90

Our implementation does not allocate.

**Pull request merged**